core/src/snapshot.rs

```diff
@@ -0,0 +1,181 @@
+const SNAPSHOT_TMP_FILE_NAME: &str = "snapshot.tmp";
+
+/// Errors produced by [`Snapshotter`] actor.
+pub type Result<T, E = ErrReport> = core::result::Result<T, E>;
+
+/// [`Snapshotter`] actor handle.
+#[derive(Clone)]
+pub struct SnapshotterHandle {
+    /// Don't used for actually sending messages but to signal that there is no more handles to [`Snapshotter`]
+    _message_sender: mpsc::Sender<()>,
+}
+
+/// Actor responsible for [`WorldStateView`] snapshot reading and writing.
+pub struct Snapshotter {
+    sumeragi: SumeragiHandle,
+    /// Frequency at which snapshot is made
+    snapshot_period: Duration,
+    /// Path to the directory where snapshots are stored
+    snapshot_path: String,
+    /// Flag to enable/disable snapshot creation
+    snapshot_creation_enabled: bool,
+    /// Flag to signal that new wsv is available for taking snapshot
+    new_wsv_available: bool,
+}
+
+impl Snapshotter {
+    /// Start [`Self`] actor.
+    pub fn start(self) -> SnapshotterHandle {
+        let (message_sender, message_receiver) = mpsc::channel(1);
+        if self.snapshot_creation_enabled {
+            tokio::task::spawn(self.run(message_receiver));
+        } else {
+            iroha_logger::info!("Snapshot creation is disabled");
+        }
+        SnapshotterHandle {
+            _message_sender: message_sender,
+        }
+    }
+
+    /// [`Self`] task.
+    async fn run(mut self, mut message_receiver: mpsc::Receiver<()>) {
+        let mut snapshot_period = tokio::time::interval(self.snapshot_period);
+        // Don't try to create snapshot more frequently if previous take longer time
+        snapshot_period.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+
+        loop {
+            tokio::select! {
+                _ = snapshot_period.tick(), if self.new_wsv_available => {
+                    // Offload snapshot creation into blocking thread
+                    self.create_snapshot().await;
+                },
+                _ = self.sumeragi.wsv_updated() => {
+                    self.new_wsv_available = true;
+                }
+                _ = message_receiver.recv() => {
+                    info!("All handler to Snapshotter are dropped. Saving latest snapshot and shutting down...");
+                    if self.new_wsv_available {
+                        self.create_snapshot().await;
+                    }
+                    break;
+                }
+            }
+            tokio::task::yield_now().await;
+        }
+    }
+
+    /// Invoke snapshot creation task
+    async fn create_snapshot(&mut self) {
+        let sumeragi = self.sumeragi.clone();
+        let path_to_snapshot = self.snapshot_path.clone();
+        let handle = tokio::task::spawn_blocking(move || -> Result<u64> {
+            sumeragi.apply_wsv(|wsv| {
+                Self::try_write_snapshot(wsv, &path_to_snapshot)?;
+                Ok(wsv.height())
+            })
+        });
+
+        match handle.await {
```

Comment:
I'm not sure this optimization actually does something.
Inside the task there is no async operations, so there is nothing `tokio` can stop at. I believe that `handle.await` will always be just `Ready` and in fact it's a still blocking operation.

Reply:
This way we allow `tokio` executor to switch to another task, [`spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) was deliberately created for this purpose (and [`Future`](https://docs.rs/tokio/latest/src/tokio/runtime/task/join.rs.html#312-344) impl `JoinHandle`).